### PR TITLE
KAFKA-14868:Remove some forgotten metrics when the replicaManager is closed

### PR DESCRIPTION
**Pull request by hudeqi**
In the ReplicaManager class, IsrExpandsPerSec, IsrShrinksPerSec, and FailedIsrUpdatesPerSec are all removed by omission.
I removed these metrics from here when it is shutdown, and added related unit tests to check to prevent forgetting to remove metrics when new metrics are added in the future.
The change has successfully passed the unit test, and when omitting to remove the metric, the unit test fails.

### CODE REVIEW

**Review by divijvaidya — COMMENTED** [review]
Thank you for the change!

Please add a unit test which verifies that the number of calls to removeMetric in this class is same as the number of call to register. This would help catch such bugs in future.

**Comment by hudeqi** [comment]
Hello, I added a count attribute to the KafkaMetricsGroup class to facilitate unit testing for all such situations. So a [jira](https://issues.apache.org/jira/browse/KAFKA-14868) was raised.
For another https://github.com/apache/kafka/pull/13473, I am going to use this ability to add corresponding unit tests after this pr is merged. @divijvaidya

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: I have an idea with which we can avoid changing the KafkaMetricsGroup. Could you please consider the following.

You can use mockito's `MockConstruction` to mock the KafkaMetricsGroup and count the number of invocations on that mock and later assert on number of invocations of add and remove. As an example, this test could be written as follows (rough example with Java code):

```
var numAddCount = 0;
var numRemoveCount = 0;
try (MockedConstruction<KafkaMetricsGroup> mockMetricsGroup = mockConstruction(KafkaMetricsGroup.class,
            (mock, context) -> {
                   doAnswer(invocation -> {
                        numAddCount++;
                    }).when(mock).newGauge(anyString());
                    
                    // similarly add mocks for newMeter etc.
                    
                    doAnswer(invocation -> {
                        numRemoveCount++;
                    }).when(mock).removeMetric(anyString());
        })) {
   val rm = new ReplicaManager(
      metrics = metrics,
      config = config,
      time = time,
      scheduler = new MockScheduler(time),
      logManager = mockLogMgr,
      quotaManagers = quotaManager,
      metadataCache = MetadataCache.zkMetadataCache(config.brokerId, config.interBrokerProtocolVersion),
      logDirFailureChannel = new LogDirFailureChannel(config.logDirs.size),
      alterPartitionManager = alterPartitionManager,
      threadNamePrefix = Option(this.getClass.getName))
      
   rm.shutdown()
   assertEquals(numAddCount, numRemoveCount)
 }
```

Comment (hudeqi) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Hello, I changed the KafkaMetricsGroup to facilitate the use of this kind of metric check for each class, and avoid mockConstruction(like above logic: numAddCount++, numRemoveCount, then compare) once for each class.
Do you think this is unnecessary? @divijvaidya

Comment (divijvaidya) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: I think that we can avoid changing the  `KafkaMetricsGroup` completely here and that is a preferred approach in my opinion because:

1. Methods in `KafkaMetricsGroup` may not be called from same thread? In that case, then we will have to use AtomicInteger for storing `numAddCount`. This will synchronize multiple threads on same lock hence, adding another point of contention.

2. Do we anticipate counters inside `KafkaMetricsGroup` to be used for any other purpose? If not, then we are adding counters only for writing unit test which is avoidable.

3. We are adding a potential source of bugs where if a new method is added is to `KafkaMetricsGroup`, the author will have to remember to correctly modify the counter.

Comment (hudeqi) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: > I think that we can avoid changing the `KafkaMetricsGroup` completely here and that is a preferred approach in my opinion because:
> 
> 1. Methods in `KafkaMetricsGroup` may not be called from same thread? In that case, then we will have to use AtomicInteger for storing `numAddCount`. This will synchronize multiple threads on same lock hence, adding another point of contention.
> 2. Do we anticipate counters inside `KafkaMetricsGroup` to be used for any other purpose? If not, then we are adding counters only for writing unit test which is avoidable.
> 3. We are adding a potential source of bugs where if a new method is added is to `KafkaMetricsGroup`, the author will have to remember to correctly modify the counter.

I see, thanks!!!

Comment (hudeqi) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:4491]: > I have an idea with which we can avoid changing the KafkaMetricsGroup. Could you please consider the following.
> 
> You can use mockito's `MockConstruction` to mock the KafkaMetricsGroup and count the number of invocations on that mock and later assert on number of invocations of add and remove. As an example, this test could be written as follows (rough example with Java code):
> 
> ```
> var numAddCount = 0;
> var numRemoveCount = 0;
> try (MockedConstruction<KafkaMetricsGroup> mockMetricsGroup = mockConstruction(KafkaMetricsGroup.class,
>             (mock, context) -> {
>                    doAnswer(invocation -> {
>                         numAddCount++;
>                     }).when(mock).newGauge(anyString());
>                     
>                     // similarly add mocks for newMeter etc.
>                     
>                     doAnswer(invocation -> {
>                         numRemoveCount++;
>                     }).when(mock).removeMetric(anyString());
>         })) {
>    val rm = new ReplicaManager(
>       metrics = metrics,
>       config = config,
>       time = time,
>       scheduler = new MockScheduler(time),
>       logManager = mockLogMgr,
>       quotaManagers = quotaManager,
>       metadataCache = MetadataCache.zkMetadataCache(config.brokerId, config.interBrokerProtocolVersion),
>       logDirFailureChannel = new LogDirFailureChannel(config.logDirs.size),
>       alterPartitionManager = alterPartitionManager,
>       threadNamePrefix = Option(this.getClass.getName))
>       
>    rm.shutdown()
>    assertEquals(numAddCount, numRemoveCount)
>  }
> ```

Sorry, I tried my best to write this unit test according to this method, but since newGauge has a return value, there is no way to use doAnswer. It seems that KafkaMetricsGroup must be changed? @divijvaidya

**Comment by divijvaidya** [comment]
@hudeqi please feel free to cherry-pick this commit: https://github.com/divijvaidya/kafka/commit/e6db8eaaeffc3b7a87f34c0e72ff25395fda97a3 into this PR.

It introduces the unit test which fails before the `removeMetrics()` and passes after that.

Edit: This was in response to your deleted comment. If you have another solution to write the unit test here, please feel free to use that. 
> Sorry, I tried my best to write this unit test according to this method, but since newGauge has a return value, there is no way to use doAnswer. It seems that KafkaMetricsGroup must be changed? [@divijvaidya](https://github.com/divijvaidya)

**Comment by hudeqi** [comment]
> @hudeqi please feel free to cherry-pick this commit: [divijvaidya@e6db8ea](https://github.com/divijvaidya/kafka/commit/e6db8eaaeffc3b7a87f34c0e72ff25395fda97a3) into this PR.
> 
> It introduces the unit test which fails before the `removeMetrics()` and passes after that.
> 
> Edit: This was in response to your deleted comment. If you have another solution to write the unit test here, please feel free to use that.
> 
> > Sorry, I tried my best to write this unit test according to this method, but since newGauge has a return value, there is no way to use doAnswer. It seems that KafkaMetricsGroup must be changed? [@divijvaidya](https://github.com/divijvaidya)

Hello, thank you for your submission, I have learned a lot, and I was not familiar with the use of mockito :)
I have merged your commit into this final commit. @divijvaidya

**Comment by divijvaidya** [comment]
I am glad that you learnt some out of this PR. Happy to help!

As next steps, in this PR, we need to ensure that the build passes to begin with. It is currently failing.
```
[2023-04-01T10:04:20.522Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-13471/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:347:5: ambiguous reference to overloaded definition,

[2023-04-01T10:04:20.522Z] both method doReturn in class Mockito of type (x$1: Any, x$2: Object*)org.mockito.stubbing.Stubber

[2023-04-01T10:04:20.522Z] and  method doReturn in class Mockito of type (x$1: Any)org.mockito.stubbing.Stubber

[2023-04-01T10:04:20.522Z] match argument types (Seq[java.io.File])
```

Could you please try running `./gradlew :core:compileTestScala` locally, fix the problems and ensure that it passes? Feel free to let me know if you need help in fixing this.

**Comment by hudeqi** [comment]
> I am glad that you learnt some out of this PR. Happy to help!
> 
> As next steps, in this PR, we need to ensure that the build passes to begin with. It is currently failing.
> 
> ```
> [2023-04-01T10:04:20.522Z] [Error] /home/jenkins/workspace/Kafka_kafka-pr_PR-13471/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:347:5: ambiguous reference to overloaded definition,
> 
> [2023-04-01T10:04:20.522Z] both method doReturn in class Mockito of type (x$1: Any, x$2: Object*)org.mockito.stubbing.Stubber
> 
> [2023-04-01T10:04:20.522Z] and  method doReturn in class Mockito of type (x$1: Any)org.mockito.stubbing.Stubber
> 
> [2023-04-01T10:04:20.522Z] match argument types (Seq[java.io.File])
> ```
> 
> Could you please try running `./gradlew :core:compileTestScala` locally, fix the problems and ensure that it passes? Feel free to let me know if you need help in fixing this.

Very strange, I successfully compiled locally, and this unit test is also OK

**Comment by hudeqi** [comment]
I try to pass in two parameters to see if it can pass the CI test.
It is also successful in my local environment, gradle version is 8.0.2, java 17, scala 2.13 @divijvaidya

**Review by divijvaidya — APPROVED** [review]
I see that the build succeeded but GitHub CI shows failed. This is not unique to this and is happening for all other PRs as well. Hence, let's wait for it to resolve.

From a code perspective, this code looks good to me! This PR is ready to be reviewed by a committer.

**Comment by hudeqi** [comment]
Hi, could you help to review this pr? @showuon

**Comment by hudeqi** [comment]
Do other committers have time to review this PR?

**Comment by hudeqi** [comment]
Hello, are you still reviewing this PR? @dajac

**Review by mimaison — APPROVED** [review]
Thanks, LGTM

**Pull request merged**
